index.ts

```diff
@@ -0,0 +1,176 @@
+import { readFileSync } from 'fs';
+
+const script = readFileSync('./incrpx.min.lua', 'utf8');
+const murmur = require('imurmurhash');
+
+/**
+ * IRedisClient is a partial interface for a Redis client. Both ioredis and
+ * node_redis are assignable to IRedisClient.
+ */
+export interface IRedisClient {
+    eval(args: any[], callback: (err: any, data: any) => void): any;
+}
+
+/**
+ * ClientProvider gets passed into the Limiter. It should be a Redis client,
+ * or a function that's called every time we want to rate limit and calls
+ * back with a Redis client.
+ */
+export type ClientProvider = IRedisClient | ((callback: (err: Error, client: IRedisClient) => void) => void);
+
+/**
+ * IRuleOptions are passed in when we want to define a rule.
+ */
+export interface IRuleOptions<T> {
+    // max number of requests to accept...
+    limit: number;
+    // in this interval (in milliseconds)
+    interval: number;
+    // hash function to use for getting IDs of the "key" passed in. Defaults
+    // to doing a murmur hash on the JSON value (or using the value itself)
+    // if it's small enough.
+    hasher?: (value: T) => string;
+}
+
+/**
+ * ILimitResult is resolved from the Limiter.limit function.
+ */
+export interface ILimitResult {
+    // limit is the total request limit in the rule.
+    limit: number;
+    // the remaining requests that can be made.
+    remaining: number;
+    // the date that the bucket expires, usually provided in an
+    // X-RateLimit-Reset header.
+    resetsAt: Date;
+    // Set to true rue if the user has exceeded their rate limit
+    isLimited: boolean;
+}
+
+export type LimitCallback = (err: Error | null, result?: ILimitResult) => void;
+
+/**
+ * Simple, default hashing function when one is not otherwise provided.
+ */
+function hasher(value: any): string {
+    if (typeof value === 'object') {
+        value = JSON.stringify(value);
+    } else {
+        value = String(value);
+    }
+
+    if (value.length > 16) {
+        value = String(murmur(value).result());
+    }
+
+    return value;
+}
+
+/**
+ * Limiter is the rate limiter parent client. You can define rules on it.
+ */
+export class Limiter {
+
+    private rules: { [name: string]: Rule<any> } = Object.create(null);
```

Comment:
Plain objects are about 3x faster than string-keyed maps. Rate limiters tend to be very hot paths, so I would tend to favor an object here.